File: aptu/302_17/tretyakova/2/src/main/java/LockFreeSet.java
import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreeSet<T extends Comparable<T>> {
    private final Node<T> tail = new Node<T>(null, null);
    private final AtomicMarkableReference<Node<T>> head = new AtomicMarkableReference<Node<T>>(tail, Node.UNMARKED);

    private static class SearchResult<T extends Comparable<T>> {
        public final AtomicMarkableReference<Node<T>> prev;
        public final Node<T> next;

        public SearchResult(AtomicMarkableReference<Node<T>> prev, Node<T> next) {
            this.prev = prev;
            this.next = next;
        }
    }

    public boolean add(T value) {
        SearchResult<T> res = search(value);
        if(res != null && res.next.getKey().equals(value)) {
            return false;
        }

        Node<T> next = head.getReference();
        Node<T> new_node = new Node<T>(value, next);

        while(!head.compareAndSet(next, new_node, Node.UNMARKED, Node.UNMARKED)) {
            next = head.getReference();
            new_node.setNext(next);
        }

        return true;
    }

    public boolean remove(T value) {
        SearchResult<T> res;
        Node<T> next;

        do {
            res = search(value);
            if (res == null || !res.next.getKey().equals(value)) {
                return false;
            }

            next = res.next.getNext();
            if(!next.isMarked()) {
                if (res.next.getAtomicNext().attemptMark(next, Node.MARKED)) {
                    break;
                }
            }
        } while(true);

        if(!res.prev.compareAndSet(res.next, next, Node.UNMARKED, Node.UNMARKED)) {
            search(res.next.getKey());
        }

        return true;
    }

    public boolean contains(T value) {
        Node<T> next = head.getReference();
        while(!next.equals(tail)) {
            if(!next.isMarked() && next.getKey().equals(value)) {
                    return true;
            }
            next = next.getNext();
        }
        return false;
    }

    public boolean isEmpty() {
        return head.getReference().equals(tail);
    }

    private SearchResult<T> search(T value) {
        AtomicMarkableReference<Node<T>> prev = head;
        Node<T> next = head.getReference();
        while(!next.equals(tail)) {
            if(!next.isMarked()) {
                if (next.getKey().equals(value)) {
                    return new SearchResult<T>(prev, next);
                } else {
                    prev = next.getAtomicNext();
                    next = next.getNext();
                }
            } else if(prev.compareAndSet(next, next.getNext(), Node.UNMARKED, Node.UNMARKED)) {
                next = prev.getReference();
            } else {
                prev = head;
                next = head.getReference();
            }
        }
        return null;
    }
}

File: hse/BPM161/nikiforovskaya/2/src/main/kotlin/ru/hse/nikiforovskaya/set/LockFreeSet.kt
package ru.hse.nikiforovskaya.set

import java.util.concurrent.atomic.AtomicMarkableReference


class LockFreeSet<T:Comparable<T>> : LockFreeSetInterface<T> {

    private val head : AtomicMarkableReference<ListNode<T>?> = AtomicMarkableReference(null, ListNode.UNMARKED)

    private class SearchResult<T : Comparable<T>>(val last: AtomicMarkableReference<ListNode<T>?>, val data: ListNode<T>?)

    override fun add(value: T): Boolean {
        val res = find(value)
        if (res.data != null && res.data.getKey() == value && !res.data.isMarked()) {
            return false
        }

        var last = res.last

        while (!last.compareAndSet(null, ListNode(value), ListNode.UNMARKED, ListNode.UNMARKED)) {

            if (last.isMarked) {
                tryAllRemoves()
            }

            val res = find(value)
            if (res.data != null && res.data.getKey() == value && !res.data.isMarked()) {
                return false
            }

            last = res.last
        }
        return true
    }

    private fun tryAllRemoves() {
        var last = head
        var node: ListNode<T>? = head.reference

        while (node != null) {
            if (node.isMarked()) {
                actuallyRemove(last, node)
            }
            last = node.next
            node = node.nextNode()
        }
    }

    override fun remove(value: T): Boolean {
        while (true) {
            val found = find(value)
            val current = found.data
            val last = found.last

            if (current != null) {
                val ref = current.nextNode()
                if (current.next.compareAndSet(ref, ref, ListNode.UNMARKED, ListNode.MARKED)) {
                    actuallyRemove(last, current)
                    return true
                }
            } else {
                return false
            }
        }
    }

    private fun actuallyRemove(last: AtomicMarkableReference<ListNode<T>?>, current: ListNode<T>) {
        var next : ListNode<T>? = current
        while (next != null && next.isMarked()) {
            next = next.nextNode()
        }
        last.compareAndSet(current, next, ListNode.UNMARKED, ListNode.UNMARKED)
    }

    override fun contains(value: T): Boolean {
        val node = find(value).data
        return node != null
    }

    override fun isEmpty(): Boolean {
        return !iterator().hasNext()
    }

    override fun iterator(): Iterator<T> {
        return getSnapshot().listIterator()
    }


    private fun find(value: T): SearchResult<T> {
        var last: AtomicMarkableReference<ListNode<T>?> = head
        var node: ListNode<T>? = head.reference

        while (node != null) {
            if (!node.isMarked() && value == node.getKey()) {
                return SearchResult(last, node)
            }
            last = node.next
            node = last.reference
        }
        return SearchResult(last, node)
    }

    private fun getSnapshot() : List<T> {
        while (true) {
            val first = getSomeList()
            val second = getSomeList()
            if (first == second) {
                return first.first
            }
        }
    }

    private fun getSomeList(): Pair<List<T>, AtomicMarkableReference<ListNode<T>?>> {
        val result = mutableListOf<T>()
        var last = head
        var node: ListNode<T>? = head.reference

        while (node != null) {
            if (!node.isMarked()) {
                result.add(node.getKey())
            }
            last = node.next
            node = node.nextNode()
        }
        return Pair(result, last)
    }
}